src/rgw/rgw_op.cc

```diff
@@ -5232,6 +5232,57 @@ int RGWCopyObj::verify_permission(optional_yield y)
 					      RGW_PERM_READ)) { 
 	  return -EACCES;
 	}
+
+  // Check for tagging permissions
+  auto tagging_drctv = s->info.env->get("HTTP_X_AMZ_TAGGING_DIRECTIVE");
+  int itag_der;
+  if (tagging_drctv) {
+    if (strcasecmp(tagging_drctv, "COPY") == 0)
+      itag_der = COPY_TAGGING_DIRECTIVE;
+    else if (strcasecmp(tagging_drctv, "REPLACE") == 0)
+      itag_der = REPLACE_TAGGING_DIRECTIVE;
+  } else {
+      itag_der = NO_TAGGING_DIRECTIVE;
+    }
+
+  int tagging_permission = 0;
+  if (Effect::Allow == eval_identity_or_session_policies(s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3PutObjectTagging,
+                                                                  obj_arn)) {
+
+    tagging_permission |= PUT_TAGGING_PERM;
+                                                                  }
+  
+  if (Effect::Allow == eval_identity_or_session_policies(s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3GetObjectTagging,
+                                                                  obj_arn)) {
+    tagging_permission |= GET_TAGGING_PERM;
+                                                                  }
+ 
+  op_ret = s->src_object->get_obj_attrs(s->obj_ctx, y, this);
+  if (op_ret == 0) {
+      attrs = s->src_object->get_attrs();
+      auto tags = attrs.find(RGW_ATTR_TAGS);
+      if (tags != attrs.end()) {
+        //tags present on source object
+        if (tagging_permission == PUT_TAGGING_PERM) {
```

Comment:
on giving more thoughts after my last comment on this, following is better way to check, since these checks would fail otherwise when both GET & PUT tagging permissions are available
>  if (tagging_permission & PUT_TAGGING_PERM)

Reply:
addressed in a different way as suggest in ceph PR https://github.com/ceph/ceph/pull/47897